segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeepertier2/StorageLedgerManager.java

```diff
@@ -0,0 +1,816 @@
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.storage.StorageNotPrimaryException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.client.BKException;
+import org.apache.bookkeeper.client.BookKeeper;
+import org.apache.bookkeeper.client.LedgerEntry;
+import org.apache.bookkeeper.client.LedgerHandle;
+import org.apache.bookkeeper.conf.ClientConfiguration;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.api.transaction.CuratorOp;
+import org.apache.curator.x.async.AsyncCuratorFramework;
+import org.apache.curator.x.async.api.DeleteOption;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+
+/**
+ * This class represents the manager for StorageLedgers.
+ * It is responsible for creating and deleting storage ledger objects in bookkeeper.
+ * This objects interacts with Zookeeper through Async curator framework to manage the metadata.
+ */
+@Slf4j
+public class StorageLedgerManager {
+    private static final BookKeeper.DigestType LEDGER_DIGEST_TYPE = BookKeeper.DigestType.MAC;
+
+    private final BookkeeperStorageConfig config;
+    private final ExecutorService executor;
+    private final AsyncCuratorFramework zkClient;
+
+    private ConcurrentHashMap<String, StorageLedger> ledgers;
+    private BookKeeper bookkeeper;
+    private long containerEpoc;
+
+    public StorageLedgerManager(BookkeeperStorageConfig config, CuratorFramework zkClient, ExecutorService executor) {
+        this.config = config;
+        this.executor = executor;
+        this.zkClient = AsyncCuratorFramework.wrap(zkClient);
+        ledgers = new ConcurrentHashMap<>();
+    }
+
+    //region Storage API
+
+    /**
+     * Creates a StorageLedger at location streamSegmentName.
+     *
+     * @param streamSegmentName Full name of the stream segment
+     * @param timeout           timeout value for the operation
+     * @return A new instance of StorageLedger if one does not already exist
+     * @Param bookkeeper the BK instance
+     * @Param zkClient
+     */
+    public CompletableFuture<LedgerData> create(String streamSegmentName, Duration timeout) {
+        StorageLedger storageLedger = new StorageLedger(this, streamSegmentName, 0, this.containerEpoc, false);
```

Comment:
Since this class requires `initialize` to be invoked prior to any other method, you should check to see if it was previously invoked. Otherwise you will get random NullPtrs or half-processed operations.

Reply:
Fixed.